spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/GatewayMetricsFilter.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2013-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.springframework.cloud.gateway.filter;
+
+import static org.springframework.cloud.gateway.support.ServerWebExchangeUtils.GATEWAY_ROUTE_ATTR;
+
+import java.util.Arrays;
+
+import org.springframework.cloud.gateway.route.Route;
+import org.springframework.core.Ordered;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.server.reactive.AbstractServerHttpResponse;
+import org.springframework.http.server.reactive.ServerHttpResponse;
+import org.springframework.web.server.ServerWebExchange;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.core.instrument.Timer.Sample;
+import reactor.core.publisher.Mono;
+
+public class GatewayMetricsFilter implements GlobalFilter, Ordered {
+
+	private MeterRegistry meterRegistry;
+
+	public GatewayMetricsFilter(MeterRegistry meterRegistry) {
+		this.meterRegistry = meterRegistry;
+	}
+
+	@Override
+	public int getOrder() {
+		return NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER - 1;
+	}
+
+	@Override
+	public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+		Sample sample = Timer.start(meterRegistry);
+		return chain.filter(exchange).then(Mono.fromRunnable(() -> {
```

Comment:
@tony-clarke-amdocs @jkschneider were talking about the `Thread.sleep` calls in the tests, is it due to this code running in a `Runnable`?  Why did you choose to implement it like this?

Reply:
@ryanjbaxter It's not related to the runnable. We do the metrics work in the mono.then() that only runs after the filter has finished. I think this is the correct place to do the work. But it means that if the client checks too soon it won't see the metric event yet.

Reply:
@spencergibb is right, using `StepVerifier` seems like it could offer a solution to the `Thread.sleep`.  I also noticed that you are using a `WebClient` in the tests rather than a `TestWebClient`.  Sorry about all the back and forth this is not an area I am too familiar with

Reply:
@ryanjbaxter @spencergibb. I don't see yet how StepVerifier helps. StepVerifer can wait for a complete signal. But the metric only gets run after the complete signal anyway. Can you provide an example of what you mean?

Reply:
I don't see how this is substantively different from `MetricsWebFilter` which doesn't use `fromRunnable` and have sleeps in its tests. Why the variation?

Reply:
The difference is when the metric gets reported. Should the metric be reported before or after the response is written to the client? Currently it is after, but I can change it to be before and then we can remove the sleep.